client/lib/service-worker/service-worker.js

```diff
@@ -100,3 +116,156 @@ self.addEventListener( 'message', function( event ) {
 			break;
 	}
 } );
+
+self.addEventListener( 'fetch', function( event ) {
+	const request = event.request;
+
+	if ( request.method !== 'GET' ) {
+		return;
+	}
+
+	// HTML Pages, fetch from the server and fallback to the Offline page
+	if ( request.mode === 'navigate' ) {
+		const previousHash = currentAssetsHash;
+		if ( navigator.onLine ) {
+			event.respondWith(
+				fetchNetworkFirst( request, OFFLINE_CALYPSO_PAGE ).then( function( response ) {
+					// Detect if the loaded page is calypso or not
+					// TODO: update nginx to reverse proxy this header
+					if ( response.headers.get( 'x-powered-by' ) === 'Express' ) {
```

Comment:
Could we set a custom header ourselves for detection instead of piggy-backing on the Express header? This seems fragile.